### PR TITLE
Kwxm/fix plc ast cbor

### PR DESCRIPTION
**Pull request by kwxm**
Fix improper use of encodeTag and also use a more space-efficient encoding for cryptographic digests.

### CODE REVIEW

**Comment by kwxm** [comment]
We'll probably want to move the Serial instance for Digest into language-plutus-core if we end up doing Merklisation because that makes extensive use of digests.

**Comment by kwxm** [comment]
Oh.  I thought this was going to be a trivial PR but think I've broken the tests.  The crowdfunding test fails, and I think it's because all the addresses are different.  I suppose this is to be expected: the AST serialisation has changed, so hashes of serialised scripts will have changed too.  I won't risk trying to fix that myself.

File: language-plutus-core/src/Language/PlutusCore/CBOR.hs [hunk 1]
```diff
@@ -44,13 +45,27 @@ However, having this flexibility allows us to encode e.g. PLC with substantial a
 for testing.
 -}
 
+
+{- Encode/decode constructor tags using encodeWord and decodeWord.  We
+   previously used encodeTag/decodeTag here, but that was wrong: those
+   are for use with a fixed set of CBOR tags with predefined meanings
+   which we shouldn't interfere with.  Note that encodeWord will only
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/CBOR.hs]: I think it's a bit confusing to phrase this note in terms of what we used to do. I do think it's helpful to have a warning about correct usage of tags for constructors - make it a proper note and maybe reference the `serialise` library.

File: plutus-wallet-api/ledger/Ledger/Orphans.hs [hunk 2]
```diff
@@ -19,13 +19,23 @@ import qualified Language.PlutusTx.Prelude  as P
 import           Schema                     (FormSchema (FormSchemaHex), ToSchema (toSchema))
 import           Type.Reflection            (Typeable)
 
+
+{- Serialising Digests from Crypto.Hash.  This is more complicated than
```
Comment (michaelpj) [inline plutus-wallet-api/ledger/Ledger/Orphans.hs]: Make this a proper note i.e. begin with `{- Note [Serialising Digetss from Crypto.Hash` on it's own line. Makes searching for them much easier!

**Comment by kwxm** [comment]
I updated the tests to use the new addresses but something's still failing.  I'm not sure what going on now.

**Comment by michaelpj** [comment]
Blame @j-mueller : that contract spec contains a hard-coded hash, which I guess you'll need to change. (I missed that during review: do we need that? It make updating the tests annoying...)

**Comment by kwxm** [comment]
I fixed that and a similar thing in TokenAccount.hs, then regenerated the various render*.txt outputs because the addresses had changed there as well.  The `plutus-use-case` tests all now pass on my laptop, but I'm getting build failures on Hydra with empty log messages.  If I run the full set of tests I get errors for `plutus-playground-server` like this:

```
 InterpreterError (CompilationErrors [RawError "Loaded package environment from /tmp/stack20034/test-ghc-env\nLoaded package environment from /tmp/stack20034/test-ghc-env",CompilationError {filename = "/tmp/playgroundrun-2c594d7b523f592b/Main.hs", row = 25, column = 1, text = [" error:","    Could not load module \8216Playground.Contract\8217","    It is a member of the hidden package \8216plutus-playground-lib-0.1.0.0\8217.","    You can run \8216:set -package plutus-playground-lib\8217 to expose it.","    (Note: this unloads all the modules in the current scope.)"
```
I'm not sure if that's what's causing the problems here.

**Comment by kwxm** [comment]
@j-mueller I see that the tests for the Future contract take much longer to run than most of the others:

```
   can initialise and obtain tokens:                             OK (21.67s)
    can increase margin:                                          OK (36.29s)
    can settle early:                                             OK (24.68s)
    can pay out:                                                  OK (24.33s)
    PIR:                                                          OK (0.48s)
    script size is reasonable:                                    Script size: 48189
```

Is that just because there are long intervals between the various events?

**Comment by michaelpj** [comment]
https://hydra.iohk.io/build/1438468#tabs-constituents

You need to fix the stylish haskell.

I don't know why the tests are slow: I have long wondered what makes a difference there, but never got around to checking.

**Comment by kwxm** [comment]
> You need to fix the stylish haskell.

Ah, thanks.  Emacs is supposed to do that for me but sometimes doesn't, for unknown reasons.

**Comment by kwxm** [comment]
It (eventually) passed the tests this time, so I'm going to quickly merge it before something else goes wrong.

**Pull request merged**